Write Rust function to compute gravitational acceleration from multip…

NeoScripter opened the pull request:
struct Vector3D {
    x: f32,
    y: f32,
    z: f32,
}

fn compute_gravitational_acceleration(displacement: Vector3D, mass: f32) -> Vector3D {
    let g: f32 = 9.81;

    let distance_squared = displacement.x.powi(2) + displacement.y.powi(2) + displacement.z.powi(2);

    if distance_squared > 0.0 {
        let magnitude = g * mass / distance_squared.sqrt();
        Vector3D {
            x: magnitude * displacement.x / distance_squared.sqrt(),
            y: magnitude * displacement.y / distance_squared.sqrt(),
            z: magnitude * displacement.z / distance_squared.sqrt(),
        }
    } else {
        Vector3D { x: 0.0, y: 0.0, z: 0.0 }
    }
}

fn compute_acceleration(masses: &[f32], positions: &[Vector3D], index_of_self: usize) -> Vector3D {
    let mut acceleration = Vector3D { x: 0.0, y: 0.0, z: 0.0 };

    for i in 0..masses.len() {
        if i != index_of_self {
            let displacement = Vector3D {
                x: positions[i].x - positions[index_of_self].x,
                y: positions[i].y - positions[index_of_self].y,
                z: positions[i].z - positions[index_of_self].z,
            };
            let grav_acceleration = compute_gravitational_acceleration(displacement, masses[i]);
            acceleration.x += grav_acceleration.x;
            acceleration.y += grav_acceleration.y;
            acceleration.z += grav_acceleration.z;
        }
    }

    acceleration
}

NeoScripter committed:
Write Rust function to compute gravitational acceleration from multiple objects

hypehuman reviewed (changes requested):
The new `compute_acceleration` function looks like it should work. But the new `compute_gravitational_acceleration` has a few issues; could you please remove that and call the existing `compute_gravitational_acceleration` function instead?

Also I think it could be simplified by using cgmath's vector instead of the new struct; that way we can do add/multiply without dealing with the individual x/y/z values.

NeoScripter committed:
Update lib.rs

Corrected version of the code

hypehuman commented on Rust/mechanics_fast/src/lib.rs:
```diff
@@ -1,12 +1,28 @@
-    const GRAVITATIONAL_CONSTANT: f64 = 6.67430e-11;
+fn compute_gravitational_acceleration(displacement: Vector3<f32>, mass: f32) -> Vector3<f32> {
+    let g: f32 = 9.81;
 
-    let distance = displacement.magnitude();
+    let distance_squared = displacement.magnitude2();
 
-    let acceleration = displacement * (GRAVITATIONAL_CONSTANT * m2) / (distance * distance * distance);
+    if distance_squared > 0.0 {
+        let magnitude = g * mass / distance_squared.sqrt();
+        displacement.normalize() * magnitude
+    } else {
+        Vector3::new(0.0, 0.0, 0.0)
+    }
+}
```
I think the original `compute_gravitational_acceleration` from d9534bdb890232396a9ac93604094880d23bef84 was perfect. The modified one is more complicated and also incorrect; it uses 9.81, which is specifically the gravitational acceleration on Earth's surface, and not applicable anywhere else. It also had the flags that allowed the C# code to call it. For reference, it was:

```
#[no_mangle]
pub extern "C" fn compute_gravitational_acceleration(displacement: Vector3<f64>, m2: f64) -> Vector3<f64> {
    const GRAVITATIONAL_CONSTANT: f64 = 6.67430e-11;

    let distance = displacement.magnitude();

    let acceleration = displacement * (GRAVITATIONAL_CONSTANT * m2) / (distance * distance * distance);

    acceleration
}
```

hypehuman commented on Rust/mechanics_fast/src/lib.rs:
```diff
@@ -1,12 +1,28 @@
+    }
+}
+
+fn compute_acceleration(masses: &[f32], positions: &[Vector3<f32>], index_of_self: usize) -> Vector3<f32> {
```
Could we make the floats 64-bit instead of 32-bit? That will make it compatible with the existing C# code.

```suggestion
fn compute_acceleration(masses: &[f64], positions: &[Vector3<f64>], index_of_self: usize) -> Vector3<f64> {
```

hypehuman committed:
Adding unit test for test_compute_gravitational_acceleration

hypehuman committed:
Adding unit test for test_compute_gravitational_acceleration

hypehuman committed:
Adding a unit test for compute_acceleration. The test doesn't build because of f32/f64 type incompatibility.

NeoScripter committed:
Update lib.rs

Reverted to the workinng version of the computer gravitational acceleration and fixed signature

hypehuman reviewed (dismissed):
Thanks, looks good!

hypehuman reviewed (approved):
Thanks, looks good!
